### PR TITLE
feat(radio): add AN/PRC-117G comms system

### PR DESCRIPTION
**Pull request by rokudara-sen**
## About the PR

This PR adds the AN/PRC-117G manpack radio for Radio Telephone Operators, a tunable direct-frequency layer for all headsets, tiered radio jamming, and an electronic-warfare loop (direction finding, interception, COMSEC changeovers) on top of it.

The goal is to make ground comms more interesting than a simple "tower works / tower does not work" check. RTOs now have a real field role: keeping squads connected, managing secure nets, handling intercepts, watching their emission signature, and reacting to jamming.

### The radio

The AN/PRC-117G is worn on the back and can only be operated by trained roles — for GOVFOR the RTO, Commander, Officer Adjutant, and Advisor; for CLF the Cell Leader. This is deliberate redundancy: if the RTO drops, another trained role can pick up the already-keyed pack and keep the element on the net. It supports up to four programmable preset slots with user-defined labels, assigned either from the faction net list or entered manually as raw frequencies. The active AN/PRC net is transmitted on with `:r`, prefixed by an operator-set callsign.

While worn and powered, the radio acts as a relay anchor. Friendly headset traffic on nets loaded into the AN/PRC works without a communications tower within 30 tiles at full signal, and out to 45 tiles with degraded signal (stock figures — see power and antennas below).

The radio has four operating modes, each with a distinct up and down so none of them is the meta:

* **FH — Frequency Hopping**: jamming against you is reduced by one tier and you can transmit from inside jam fields; costs 1.5× battery and even secured FH traffic carries a small DF signature. Interceptors without matching fill get light hop interference on in-the-clear FH traffic.
* **SC — Single Channel**: the efficient workhorse. 0.75× battery and zero DF signature while secured; no jamming resistance.
* **CT — Cipher Text**: invisible to interceptors (no red text, no log entry) and no DF signature, but only fill-loaded AN/PRCs can decode it — your own riflemen hear an encrypted burst, so it is useless for squad broadcast. Transmit is locked until fill is loaded. No jamming resistance.
* **PT — Plain Text**: the emergency mode. 0.5× battery, but readable by everyone, jammed one tier harder, and the biggest DF signature.

It also has monitor (receive-only) mode, scan mode with automatic active-slot switching, a `0–4` squelch noise gate, and a 50-entry server-side net log that survives closing the panel and powering off. The log can be filtered client-side by net (dropdown) and by free-text search over sender and message.

A **TX power** dial (LO/MED/HI) trades reach against battery drain and DF signature: LO is 0.6× range at half battery and half signature, HI is 1.5× range at double both.

The radio runs on a swappable battery (standby drain plus a larger draw per transmission or radio check; auto power-off when depleted) and has an antenna slot: the stock **whip** holds range steady on the move, while the **mast** roughly doubles range when stationary but drops to about *half* stock range while moving — worse than the whip, so neither antenna dominates. A **RADIO CHECK** button transmits a standard check on the active net and privately reports which stations are in clear or degraded range; run it inside jamming and it also reports a compass bearing to the strongest emitter, making jammer hunting possible.

### Retrans stations

A radio does not have to be worn to relay. An RTO can plant it as an unattended retrans station: it anchors its loaded faction nets, keeps logging traffic, counts as stationary for a mast antenna, and drains its battery until packed up. *Anyone* — including the enemy — can pack up a planted station, which hands them a keyed radio and its fill card. That is the capture mechanic.

### COMSEC

COMSEC is handled through physical fill cards with a dedicated slot. Faction nets are encrypted end to end: an interceptor tuned to your frequency hears *all* headset traffic as static unless they hold a matching current-generation fill card. The AN/PRC itself needs fill loaded to join that encryption — transmitting on a faction net without it warns the operator and goes out in the clear. **ZEROIZE** ejects the card, **DESTROY** deletes it when there is no time to run.

The fill slot accepts captured enemy cards, which is how hostile secured traffic (including CT and the FH hopset) is decoded. The counter is the **RECRYPTO** order: with a current card of your own faction loaded, it supersedes every fill card issued before it — including whatever the enemy captured — turning them into identifiable but non-securing stale fill. Replacement cards are not conjured by the order; freshly spawned cards stamp themselves with the current generation, so resupply through the normal supply chain always arrives valid.

### Direction finding

Transmitting with a DF signature carries a chance to reveal the sender's approximate position as a temporary contact blip on the opposing faction's tactical map. PT is the brightest beacon, unsecured faction-net traffic is nearly as bad, secured FH radiates a small signature, and secured SC/CT have none. The chance scales with TX power, rises near active enemy jamming, and *accumulates* for repeated in-the-clear transmissions from the same spot — going quiet or displacing resets it. Transmit, then move.

### Headsets and jamming

All headsets also get one tunable direct frequency, from `30.000` to `87.999 MHz`, transmitted with `:x`. Direct frequencies work peer-to-peer without towers, keys, or relays, and can be bridged over distance by an AN/PRC carrying the same frequency.

Radio jammers now interact with all of this. Standard radios are still hard-blocked when transmitting from inside a jam field, but AN/PRC transmissions degrade instead with mode-dependent garble; receiver proximity to jammers applies additional garbling, degraded messages always corrupt at least one word, and direct frequencies can be jammed too.

### Content

GOVFOR, OPFOR, and CLF all get their own AN/PRC variants, fill cards, and antennas. The trained roles above carry their faction's fill card; the RTO is issued a pre-keyed radio, a spare fill card, and a laminated frequency assignment card. Resupply (radios, batteries, fill, antennas) goes through requisitions/ASRS for GOVFOR and OPFOR and faction vendors for CLF, including a resupply crate.

Three guidebook pages under a Communications Reference index cover the AN/PRC-117G, field headsets, and a one-page quick reference (prefixes, ranges, mode/power/squelch tables) for mid-round lookups.

## Why / Balance

This gives the RTO role something important to actively manage during a round. Squads benefit from staying near their RTO, losing the RTO can hurt a whole element's communications, and capturing a radio, retrans station, or fill card provides real intelligence value. (This could also be used as an intel objective.)

The COMSEC loop (fill, intercept, zeroize, capture, recrypto) gives faction RTOs something to play around without changing how regular soldiers use their normal headset encryption keys.

Every dial is a tradeoff on purpose, so there is no single correct configuration:

* Modes: FH pays battery and a DF signature for jam resistance; SC is quiet and cheap but folds under jamming; CT is unreadable and undetectable but only talks to other filled AN/PRCs; PT is the cheap emergency mode everyone can read.
* TX power: reach costs battery and DF exposure in both directions.
* Antennas: the mast doubles range standing still but is *worse* than the whip on the move, so it belongs at command posts and retrans sites, not on patrol.
* DF accumulation punishes transmitting repeatedly from one spot in the clear, rewarding real signals discipline.

The scope is intentionally contained. The relay only ever carries the operator faction's own nets. Colony and third-party channels still route the way they did before: tower-dependent and unaffected by an RTO simply tuning to them.

Colonist and third-party gameplay should remain unchanged unless someone brings a jammer into play, which affects everyone equally.

## Please test this yourself

Please do not just review the code and assume this is fine. This touches a lot of radio behavior, and I would really appreciate people testing it ingame with weird setups.

Good things to try:

* [ ] Use two or more RTOs on the same faction net.
* [ ] Put different faction RTOs close to each other and check what does and does not relay.
* [ ] Try using the AN/PRC while moving between grids, maps, dropships, shuttles, or teleports.
* [ ] Kill, crit, cuff, strip, disconnect, or move the RTO and check whether relay behavior stops correctly.
* [ ] Put the radio in containers, drop it, turn it off, delete it, or move it away from the operator.
* [ ] Try friendly fill, enemy fill, no fill, zeroized fill, and stale (recrypto'd) fill.
* [ ] Order a recrypto and confirm captured cards stop decoding while fresh supply arrives current.
* [ ] Put jammers near the sender, receiver, relay, or all of them at once.
* [ ] Transmit repeatedly from one spot in PT near jamming and watch the enemy tactical map for DF blips; then displace and confirm the accumulation resets.
* [ ] Plant a retrans station, walk out of worn-relay range, and confirm it carries the net; pack it up as the enemy faction.
* [ ] Swap whip/mast antennas while moving and standing still, at each TX power, and confirm the range changes (radio check helps here).
* [ ] Drain a battery to zero mid-conversation and hot-swap it.
* [ ] Try direct-frequency headset comms between hostile groups using the same frequency.
* [ ] Check implants, accessory earpieces, headset encryption keys, muted channels, ghosts, and admin observing.
* [ ] Fill the net log past 50 entries with several nets loaded, then filter by net and by text with the panel open while traffic keeps arriving.
* [ ] Spam several relays, several jammers, and many different direct frequencies to see if anything breaks.

If you can think of an edge case that is not listed here, please reply with it. I would rather have too many strange test cases than miss one obvious exploit.

## Ingame testing performed

The following has already been tested ingame:

* [x] AN/PRC-117G can be equipped on the back slot.
* [x] Only trained operators can operate the AN/PRC-117G.
* [x] Radio presets can be configured from faction nets.
* [x] Radio presets can be configured through raw frequency entry.
* [x] `:r` transmits on the selected active AN/PRC net.
* [x] Friendly headset traffic is relayed by a worn, powered AN/PRC.
* [x] Relay coverage works without a communications tower within the intended relay range.
* [x] Tower-gating exemptions work for relay-covered traffic.
* [x] Relay-covered traffic degrades outside the full-signal range.
* [x] The relay only carries the operator faction's own nets.
* [x] Tuning enemy, colony, or third-party frequencies does not reroute those channels for uninvolved players.
* [x] Direct headset frequency tuning works within the `30.000–87.999 MHz` range.
* [x] `:x` transmits on the headset's tuned direct frequency.
* [x] Direct-frequency headset traffic works peer-to-peer without towers.
* [x] AN/PRC bridging works for headsets using the same direct frequency.
* [x] Matching COMSEC fill allows secured faction-net traffic to be decoded.
* [x] Missing or incorrect COMSEC fill turns secured AN/PRC traffic into static.
* [x] Captured enemy fill cards can be inserted and used to decode hostile secured traffic.
* [x] Enemy-net intercept traffic appears as red intercept text.
* [x] ZEROIZE ejects the inserted fill card.
* [x] Jammers hard-block standard radio transmissions from inside a jam field.
* [x] Jammers degrade AN/PRC transmissions instead of hard-blocking them.
* [x] FH and PT modes apply their intended jamming resistance/susceptibility behavior.
* [x] Direct frequencies are affected by jamming.
* [x] GOVFOR and CLF variants spawn with their expected faction equipment.
* [x] Guidebook pages open and display correctly.

Newer systems — tick these off as testing is completed:

* [x] DESTROY deletes the loaded fill card.
* [x] RECRYPTO supersedes older cards faction-wide; stale cards stop securing and identify as STALE; enemy cards cannot trigger a recrypto for your faction.
* [x] TX power scales relay range, battery cost, and DF exposure at each setting.
* [x] DF blips appear on the opposing faction's tactical map for in-the-clear/PT traffic near jamming, and accumulation resets on displacement or going quiet.
* [x] Retrans plant/pack-up flow, including enemy capture and mast-on-retrans range.
* [ ] Radio check coverage report, including the jammer bearing while inside a jam field.
* [ ] Battery drain on standby vs. per transmission, auto power-off, and hot-swap.
* [ ] Whip vs. mast antenna ranges while moving and stationary.
* [ ] Net log filters (per-net dropdown, text search, CLR) with live traffic arriving.

## Outstanding before merge

* [ ] OPFOR loadout wiring still needs to be completed/verified.
* [x] Media/screenshots for the PR.

## Merge warning

> [!WARNING]
> Do **not** merge this PR until it has been tested heavily
>
> This touches radio routing, tower gating, jamming, faction encryption, the tactical map, item slots, headset behavior, and intercept logic. There are probably edge cases I have not thought of yet.

## Edge cases reviewers should think about

Please reply to this PR with any abuse cases, edge cases, or strange interaction ideas you can think of, especially around:

* [ ] Multiple RTOs relaying the same net on the same map.
* [ ] Multiple RTOs from different factions relaying near each other.
* [ ] AN/PRC relay behavior across grids, maps, shuttles, dropships, and FTL/teleport movement.
* [ ] Radio behavior when the RTO dies, disconnects, is crit, is cuffed, is stripped, or drops the radio.
* [ ] Relay behavior when the AN/PRC is powered off, unworn, deleted, gibbed, EMPed, or moved into containers.
* [ ] Captured radios with friendly fill, enemy fill, no fill, zeroized fill, or stale fill.
* [ ] Recrypto abuse: ordering changeovers with captured or stale cards, recrypto spam, and cards spawned mid-changeover.
* [ ] Retrans stations: planting in odd places, capture griefing, planted radios on moving grids, and mast range while planted.
* [ ] DF blips: correct faction targeting, blip cleanup after expiry, and interaction with tactical map snapshots for non-live viewers.
* [ ] Ways to cheese the mast antenna's stationary check (stutter-stepping, being pulled or carried, standing on moving grids/conveyors).
* [ ] Jamming behavior when the transmitter, receiver, relay, or all three are inside different jam fields.
* [ ] Direct-frequency abuse cases, especially shared frequencies between hostile groups.
* [ ] Interaction with implants, accessory earpieces, headset encryption keys, channel mutes, and ghost/admin observing.
* [ ] UI desyncs between the radio panel, headset tuning window, item slots, net log, and net log filters.
* [ ] Roundstart loadouts, latejoin loadouts, role changes, respawns, and admin-spawned equipment.
* [ ] Performance with many players, many direct frequencies, several jammers, and multiple active relays.
* [ ] System ordering: `ANPRCRangeSystem` relies on `after:` constraints against `RMCTelephoneSystem` / `JammerSystem`. If upstream ever renames those systems, this fails loudly at server startup rather than silently.
* [ ] Engine subscription constraints: only one system may subscribe a given `(component, event)` pair. If upstream ever subscribes `<HeadsetComponent, MapInitEvent>` itself, the server should fail loudly at startup with a `Duplicate Subscriptions` exception instead of silently changing behavior.

## Technical details

System layout:

* `ANPRCRadioSystem` (split into core / `.Transmit.cs` / `.Bui.cs` partials): presets, transmit, receive/intercept, DF exposure, retrans, and net log.
* `ANPRCCryptoSystem`: fill card item slot, fill generations, and recrypto changeovers.
* `ANPRCGarbleSystem`: jam, range, COMSEC garble tiers, and squelch. Jam intensity is cached per tick.
* `ANPRCRangeSystem`: per-map relay range gating, antenna/TX-power effective ranges.
* `TunableFrequencySystem`: direct-frequency routing and relay bridging.
* `TacticalMapSystem.AU14.cs`: DF contact blips (written to both the live and snapshot blip dictionaries, since non-live viewers only render the snapshot).

Notes:

* `ANPRCRangeSystem` subscribes to `RadioSendAttemptEvent` / `RadioReceiveAttemptEvent` **after** `RMCTelephoneSystem` and `JammerSystem`, then overrides their verdict for relay-covered traffic and AN/PRC `:r` transmissions. Conditions that must stay cancelled, such as muted channels via `RMCRadioFilter` and jammed non-ANPRC senders, are checked again before anything is un-cancelled.
* `ANPRCGarbleSystem` subscribes to `RadioReceiveEvent` on `TunableHeadsetComponent` **before** `HeadsetSystem`. The by-ref event carries the garbled message into the normal relay-to-chat path.
* `TunableFrequencySystem` injects `TunableHeadsetComponent` and registers its BUI on every headset at map-init through `SharedUserInterfaceSystem.SetUi()`, instead of editing the upstream `CMHeadset` prototype.
* **Engine constraint:** only one system may subscribe a given `(component, event)` pair. Our `_AU14` systems therefore hook `TunableHeadsetComponent` for receive-garbling and `<HeadsetComponent, MapInitEvent>` for component injection. If upstream ever subscribes `<HeadsetComponent, MapInitEvent>` itself, the server will probably fail at startup with a `Duplicate Subscriptions` exception; the fix would be retargeting our subscription.
* The radio item carries its own `ActiveRadioComponent`, like a headset, instead of granting channels to the wearer. This avoids collisions with accessory earpieces or radio implants that put `IntrinsicRadioReceiverComponent` on the player.
* Mode and TX-power multipliers (battery, range, DF) live in shared `RadioModeExtensions` / `RadioTxPowerExtensions`, so client UI and server logic can't drift apart.
* Antenna ranges are copied from the seated antenna item onto the relay anchor; a stationary-only antenna whose wearer is moving collapses to stock range × its `MovingRangeMultiplier` (0.5 for the mast). Planted retrans stations always count as stationary.
* The net log is server-side (50 entries, survives UI close and power-off); the net dropdown and text-search filters are purely client-side view filtering in the radio window and never touch the recorded log.
* Added a new BUI and XAML window for the radio panel, and a small frequency-tuning window for headsets.
* Added prototypes for:

  * Faction-specific AN/PRC variants, batteries, and whip/mast antennas.
  * Faction fill cards and the frequency assignment card.
  * A resupply crate plus requisitions catalog and vendor entries.
  * Sentinel radio channels for `:r` and `:x` routing.
* Added guidebook XML (Communications Reference index, AN/PRC-117G, Field Headsets, Comms Quick Reference) and guide entries.
* Added FTL locale files.

## Media

image
image
image
image
image
image
image
image
image
image

## Requirements

* [x] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
* [x] I have added media to this PR or it does not require an ingame showcase.
* [x] By submitting this code and/or assets, I confirm that I either own them or have provided the correct necessary licenses to use and distribute them. I agree to be fully responsible for any legal claims or issues arising from the use of these materials.

**Changelog**

:cl:

* add: Added the AN/PRC-117G manpack radio for Radio Telephone Operators and command roles, with up to four programmable preset nets, operator callsigns, and `:r` transmission.
* add: Added relay-based ground comms. Headset traffic on nets loaded in a worn AN/PRC works without a comms tower within 30 tiles at full signal and 45 tiles with degraded signal.
* add: Added AN/PRC operating modes: Frequency Hopping, Single Channel, Cipher Text, and Plain Text, each trading battery cost, jamming resistance, security, and direction-finding signature so no single mode dominates.
* add: Added a transmit power dial (LO/MED/HI) to the AN/PRC, trading relay range against battery drain and direction-finding exposure.
* add: Added physical COMSEC fill cards for GOVFOR, OPFOR, and CLF. Secured faction-net traffic is pure static to interceptors without the matching fill, and captured enemy cards can be loaded to decode hostile traffic.
* add: Added the RECRYPTO order: a faction-wide COMSEC changeover that supersedes all previously issued fill cards, including captured ones. Fresh fill arrives through the normal supply chain already current.
* add: Added direction finding. Transmitting in the clear, in PT mode, or near enemy jamming can reveal the sender's position as a temporary contact blip on the opposing faction's tactical map, with repeated transmissions from the same spot raising the odds.
* add: Added retrans stations. A planted AN/PRC relays its faction nets unattended, logs traffic, and drains its battery — and anyone, including the enemy, can pack it up and capture it.
* add: Added signal interception. Tuning an enemy net's frequency on the AN/PRC delivers their traffic as red intercept text and records it in the net log.
* add: Added monitor mode, scan mode, a squelch noise gate, and a filterable 50-entry net log to the AN/PRC.
* add: Added a battery slot and an antenna slot to the AN/PRC. The stock whip antenna holds relay range steady on the move; the mast antenna roughly doubles range while stationary but reaches only about half of stock range while moving.
* add: Added a radio check action that reports which stations on the net are in clear or degraded range — and, inside jamming, a bearing to the strongest emitter.
* add: Added a tunable direct frequency, from `30.000` to `87.999 MHz`, to all headsets. It transmits with `:x`, requires no tower or encryption key, and can be bridged by an AN/PRC.
* add: Added GOVFOR, OPFOR, and CLF AN/PRC radio variants, fill cards, and antennas, with faction loadout wiring where currently implemented, plus resupply through requisitions and faction vendors.
* add: Added guidebook pages for the AN/PRC-117G, field headsets, and a comms quick reference under the Communications Reference.
* tweak: Radio jammers now degrade AN/PRC transmissions with mode-dependent garble instead of blocking them, jam both endpoints of a transmission, and affect direct frequencies.
* tweak: Transmitting on a faction net without COMSEC fill now warns the operator on every transmission.
* tweak: RTOs broadcasting in Frequency Hopping mode cause light interference for enemy interceptors without matching fill.

### CODE REVIEW

File: Resources/Prototypes/_AU14/Roles/Jobs/Military/Base/commanderBase.yml [hunk 1]
```diff
@@ -73,7 +73,9 @@
     ears: AU14HeadsetGovforCommand
     pocket1: AU14StampCO
   inhand:
-    -  AU14TabletGovfor
+    - AU14TabletGovfor
+    - ANPRCFillCardGOVFOR
+    - ANPRCFillCardGOVFOR
```
Comment (maertayn) [inline Resources/Prototypes/_AU14/Roles/Jobs/Military/Base/commanderBase.yml]: The rank chevron is spawned and equipped into a free hand. Current implementation causes the Captain's chevron to be yoinked by anyone that spawns next to them. 

Move this to pocket2 or the back storage (marines spawn with a satchel which isn't listed here)

File: Resources/Prototypes/_AU14/Entities/Objects/anprc117g.yml [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    interfaces:
+      enum.ANPRCRadioUI.Key:
+        type: ANPRCRadioBoundUserInterface
+
+  - type: Appearance
+
+  - type: ClothingSpeedModifier
+    walkModifier: 0.95
+    sprintModifier: 0.95
+
```
Comment (maertayn) [inline Resources/Prototypes/_AU14/Entities/Objects/anprc117g.yml]: What's with all the newlines here in this prototype?

**Comment by troytroy400** [comment]
Media is gonna be required for this one when you get the chance. Otherwise this looks really cool, RTO can probably make a comeback as a dedicated role once more.

File: Content.Server/_AU14/Radio/ANPRCRadioSystem.cs [hunk 3]
```diff
@@ -0,0 +1,765 @@
...
+
+        var outMessage = string.IsNullOrWhiteSpace(radio.Callsign)
+            ? args.Message
+            : $"{radio.Callsign} {args.Message}";
+
+        // Raw-frequency slots route peer-to-peer instead of through RadioSystem.
+        if (radio.FrequencyOverrides.TryGetValue(radio.ActiveSlot, out var dynFreq))
+        {
+            args.Channel = null;
+            _tunable.BroadcastOnFrequency(args.Source, dynFreq, outMessage);
```
Comment (TheHellFireo) [inline Content.Server/_AU14/Radio/ANPRCRadioSystem.cs]: This in not good. you are just making a new chat system which has none of the systems the current one has like admin logging, metadata, observer checks etc etc.

File: Content.Server/_AU14/Radio/TunableFrequencySystem.cs [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+            if ((receiverPos - anprcPos).Length() > PartialRange)
+                continue;
+
+            return anprcUid;
+        }
+
+        return null;
+    }
+
+    private void SendToEntity(EntityUid sender, EntityUid receiver, string message, int frequency)
```
Comment (TheHellFireo) [inline Content.Server/_AU14/Radio/TunableFrequencySystem.cs:348]: Same here

File: Content.Server/_AU14/Radio/TunableFrequencySystem.cs [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+
+    private void OnUnequipped(Entity<TunableHeadsetComponent> ent, ref GotUnequippedEvent args)
+    {
+        if (TryComp(args.Equipee, out TunedFrequencyComponent? wearer) && wearer.Source == ent.Owner)
+            RemComp<TunedFrequencyComponent>(args.Equipee);
+    }
+
+    // :x intercept
+
+    private void OnSpoke(Entity<TunedFrequencyComponent> ent, ref EntitySpokeEvent args)
```
Comment (TheHellFireo) [inline Content.Server/_AU14/Radio/TunableFrequencySystem.cs:88]: This only checks for the radio component and doesn't checkj if the headset is broken or disabled etc

File: Content.Server/_AU14/Radio/ANPRCRadioSystem.cs [hunk 6]
```diff
@@ -0,0 +1,765 @@
...
+    private void UpdateEquippedChannels(Entity<ANPRCRadioComponent> ent)
+    {
+        RevokeReceiveChannels(ent);
+        GrantReceiveChannels(ent);
+    }
+
+    private void UpdateBuiState(Entity<ANPRCRadioComponent> ent)
+    {
+        _ui.SetUiState(ent.Owner, ANPRCRadioUI.Key, new ANPRCRadioState(
+            new Dictionary<int, ProtoId<RadioChannelPrototype>>(ent.Comp.Presets),
```
Comment (TheHellFireo) [inline Content.Server/_AU14/Radio/ANPRCRadioSystem.cs]: You rebuild the UI even when its not open

File: Content.Server/_AU14/Radio/ANPRCRangeSystem.cs [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+        if (HasComp<ANPRCRadioComponent>(args.RadioSource))
+        {
+            args.Cancelled = false;
+            return;
+        }
+
+        var channelId = args.Channel.ID;
+
+        // Same map-local rule as OnSendAttempt.
+        if (!AnyAnchorCoversChannel(channelId, Transform(args.RadioReceiver).MapID))
```
Comment (TheHellFireo) [inline Content.Server/_AU14/Radio/ANPRCRangeSystem.cs]: You already scan for anchors in line 127, this just doubles the work

**Review by TheHellFireo — COMMENTED** [review]
Other then stuff i found, i really don't know how this would be good for gameplay
This could have just been a fun little minigame for RTO, where they exist as a limited ranged radio tower ie
,
 if the radio tower is down on the map, the RTO has to adjust signals and manage channels every few odd minutes to be able to have players in like a 15 tile radius be able to hear and transmit radio messages like normally

**Comment by maertayn** [comment]
Dropped a banger and then went completely dark

**Comment by MACMAN2003** [comment]
oh my god... it's beautiful...

**Comment by troytroy400** [comment]
I am gonna ask that you make an in-game guide post alongside this PR

**Comment by rokudara-sen** [comment]
> I am gonna ask that you make an in-game guide post alongside this PR

Already added

**Comment by rokudara-sen** [comment]
Since this PR is massive, I have decided to split it up into multiple PRs. I will close this one.
